File: 0937-reorder-data-in-log-files/0937-reorder-data-in-log-files.java
class Solution {
    public String[] reorderLogFiles(String[] logs) {
        // 문자 로그를 지정할 문자 리스트
        List<String> letterList = new ArrayList<>();

        // 숫자 로그를 지정할 숫자 리스트
        List<String> digitList = new ArrayList<>();

        for(String log : logs) {
            if(Character.isDigit(log.split(" ")[1].charAt(0))) {
                digitList.add(log);
            } else {
                letterList.add(log);
            }
        }
               
               
        // 문자 리스트 정렬
        letterList.sort((s1, s2) -> {
                String[] s1x = s1.split(" ", 2);
                String[] s2x = s2.split(" ", 2);

                // 식별자 외 나머지 부분 : 문자 로그 사전순 비교
                int compared = s1x[1].compareTo(s2x[1]);

                // 식별자 : 문자가 동일한 경우 식별자 비교
                if(compared == 0) {
                    return s1x[0].compareTo(s2x[0]);
                } else {
                    /*
                    * s1과 s2가 동일한 경우 0 반환
                    * s2가 앞으로 와야한다면 양수(1) 반환
                    * s2가 뒤에 머물러야 한다면 음수(-1)을 반환한다.
                    */
                    return compared;
                }
        });
        
        // 문자 리스트와 숫자 리스트 합치기, 조건 상 문자 리스트 뒤에 숫자 리스트가 온다.
        // 숫자 로그는 '입력 순서대로 한다'라는 조건이 있었으므로 따로 정렬하지 않는다.
        letterList.addAll(digitList);
        
        // 리스트를 String 배열로 변환해 리턴 (toArray())
        return letterList.toArray(new String[0]);

    }
}

File: 0937-reorder-data-in-log-files/0937-reorder-data-in-log-files.kt
class Solution {
    fun reorderLogFiles(logs: Array<String>): Array<String> {
        // 문자 로그를 저장할 문자 리스트
        val letterList = mutableListOf<String>()
        
        // 숫자 로그를 저장할 숫자 리스트
        val digitList = mutableListOf<String>()
        
        for(log in logs) {
            if(Character.isDigit(log.split(" ")[1][0])) {
                digitList.add(log)
            } else {
                letterList.add(log)
            }
        }
        
        // 문자 리스트 정렬
        letterList.sortWith(Comparator { s1: String, s2: String ->
            // 식별자와 식별자 외 나머지 부분을 분리
            val s1x = s1.split(" ", limit = 2)
            val s2x = s2.split(" ", limit = 2)
            
            // 문자 로그 사전순 비교
            val compared = s1x[1].compareTo(s2x[1])
            if(compared == 0) {
                // 문자가 같다면 식별자를 비교
                s1x[0].compareTo(s2x[0])
            } else {
                // s1x[1]과 s2x[1]이 같다면 0, s1x[1]이 뒤에 와야한다면 1, s1x[1]가 앞에 머무른다면 -1을 반환
                compared
            }
        })
        
        // 문자 리스트 뒤로 숫자 리스트를 붙인다.
        // 숫자 로그는 '입력 순서대로'라는 제약조건이 있으므로 따로 정렬하지 않는다.
        letterList.addAll(digitList)
        
        // toTypedArray() : 리스트의 요소들을 동일한 타입의 배열로 변환하여 반환
        return letterList.toTypedArray()
    }
}